docs/contributing/provider-design.md

```diff
@@ -0,0 +1,127 @@
+
+* Batch Job Queues
+* CloudFront Distributions
+* RDS DB Event Subscriptions
+
+It is recommended that this ability be implemented within the parent Terraform resource instead of creating a separate resource.  Trying to manage updates that require the resource to be in a specific running state becomes impractical for operators to manage within a Terraform configuration. This is generally a consistency and future-proofing design decision, even if there is no problematic resource update use case within the current API.
+
+### Task Execution and Waiter Resources
+
+Certain AWS functionality operates with the execution of individual tasks or external validation processes that have a completion status. Some examples of these include:
+
+* ACM Certificate validation
+* EC2 AMI copying
+* RDS DB Cluster Snapshot management
+
+The general expectation in these cases is that the AWS service APIs provide sufficient lifecycle management to start the execution/validation and read the status of that operation. At the end of the operation, the result may be something that can be fully managed on its own with update and delete support (e.g. with EC2 AMI copying, a copy of the AMI is a manageable AMI). In these cases, the preference is to create a separate Terraform resource, rather than implement the functionality within the parent Terraform resource. This recommendation is even at the expense of duplicated resource and schema handling.
+
+One related consideration is the ability to manage the running state of a resource (e.g. start/stop it). See the [Managing Resource Running State section](#managing-resource-running-state) for specific details.
+
+## Other Considerations
+
+### AWS Credential Exfiltration
+
+In the security-minded interest of the community, the provider will not implement data sources that will give operators the ability to reference or export the AWS credentials of the running provider. While there are valid use cases of this information, such as using those AWS credentials to execute AWS CLI calls as part of the same Terraform configuration, it also becomes possible for those credentials to be discovered and used completely outside of the current Terraform execution. Some of the concerns around this topic include:
+
+* The values can potentially be visible in Terraform user interface output or logging where anyone with access to those can see and use them.
+* The values must be stored in the Terraform state in plaintext, due to how Terraform operates. Anyone with access or another Terraform configuration that references the state can see and use them.
+* Any new functionality related to this, while opt-in to implement, is also opt-in to prevent via security controls or policies. Since this introduces a weak by default security posture, organizations wishing to implement those controls may require advance notice before the release or may never be able to update.
```

Comment:
Minor tweaks

```suggestion
In the interest of security, the maintainers will not approve data sources that provide the ability to reference or export the AWS credentials of the running provider. There are valid use cases for this information, such as to execute AWS CLI calls as part of the same Terraform configuration. However, this mechanism may allow credentials to be discovered and used outside of Terraform. Some specific concerns include:

* The values may be visible in Terraform user interface output or logging, allowing anyone with user interface or log access to see the credentials.
* The values are currently stored in plaintext in the Terraform state, allowing anyone with access to the state file or another Terraform configuration that references the state access to the credentials.
* Any new related functionality, while opt-in to implement, is also opt-in to prevent via security controls or policies. Adopting a weaker default security posture requires advance notice and prevents organizations that implement those controls from updating to a version with any such functionality.
```